utils/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum State {
+    Ongoing(usize),
+    Finished(usize),
+    Empty,
+}
+
+/// # Examples
+///
+/// Compile time intersection assert.
+/// Will panic! in case duplicated messages were provided.
+/// Requires sorted arrays to work.
+/// ```
+///     const _: () = {
+///         let msgs: [&[&str]; 2] = [&["msg_a", "msg_b"], &["msg_c", "msg_d"]];
+///         utils::assert_no_intersection(msgs);
+///     };
+/// ```
+pub const fn assert_no_intersection<const N: usize>(msgs: [&[&str]; N]) {
```

Comment:
All those functions seems correctish, but are super-complicated. They deserve proper non-doc comments about how and why they do what they do and unit tests for every single of them, not only the entry point.